src/Microsoft.OData.Core/Value/ODataResourceValue.cs

```diff
@@ -0,0 +1,38 @@
+//---------------------------------------------------------------------
+// <copyright file="ODataResourceValue.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+
+namespace Microsoft.OData
+{
+    /// <summary>
+    /// Represents the value of a resource (complex or entity).
+    /// </summary>
+    public sealed class ODataResourceValue : ODataValue
+    {
+        /// <summary>
+        /// Gets or sets the type name.
+        /// </summary>
+        public string TypeName { get; set; }
+
+        /// <summary>
+        /// Gets or sets the properties belong to this resource.
+        /// </summary>
+        public IEnumerable<ODataProperty> Properties { get; set; }
+
+        /// <summary>
+        /// Collection of custom instance annotations.
+        /// </summary>
+        [SuppressMessage("Microsoft.Usage", "CA2227:CollectionPropertiesShouldBeReadOnly",
+            Justification = "We want to allow the same instance annotation collection instance to be shared across ODataLib OM instances.")]
+        public ICollection<ODataInstanceAnnotation> InstanceAnnotations
+        {
+            get { return this.GetInstanceAnnotations(); }
+            set { this.SetInstanceAnnotations(value); }
+        }

```

Comment:
Seems inconsistent that I have instance annotations on ODataResourceValue and not other ODataValues.  I understand why we wouldn't want to add annotations to all of the ODataValues (since, in the non-annotation case, the annotations are applied to the property), but how would I write an annotation on a primitive value used in an annotation?

Reply:
Discussed it offline.
So far, ODL doesn't allow to annotation on ODataPrimitiveValue or other ODataValue. 
However, the discussion is open. We should discuss it more later. And I filed an issue at https://github.com/OData/odata.net/issues/1353 to track on it.